s3/utils.bal

```diff
@@ -407,3 +423,43 @@ isolated function handleHttpResponse(http:Response httpResponse) returns @tainte
         return error(xmlPayload.toString());
     }
 }
+
+isolated function getIAMCredentials() returns IAMCredentials|error {
+    // Create HTTP client for metadata service
+    http:Client metadataClient = check new (METADATA_TOKEN_URL);
+
+    // Get IMDSv2 token
+    http:Response tokenResponse = check metadataClient->put("", {}, {[X_AWS_EC2_METADATA_TOKEN_TTL_SECONDS]: "21600"});
+    string token = check tokenResponse.getTextPayload();
+
+    // Get role name
+    http:Client roleClient = check new (METADATA_BASE_URL);
+    http:Response roleResponse = check roleClient->get("",
+        headers = {
+        [X_AWS_EC2_METADATA_TOKEN]: token
+    }
+    );
+    string roleName = check roleResponse.getTextPayload();
+
+    // Get credentials
+    http:Client credClient = check new (METADATA_BASE_URL + "/" + roleName);
+    http:Response credResponse = check credClient->get("",
+        headers = {
+        [X_AWS_EC2_METADATA_TOKEN]: token
+    }
+    );
+
+    // Parse JSON payload with full error handling
+    json|error jsonResponse = credResponse.getJsonPayload();
+    if jsonResponse is error {
+        return jsonResponse;
+    }
+
+    // Detailed type conversion with error handling
+    IAMCredentials|error credsResult = jsonResponse.cloneWithType(IAMCredentials);
+    if credsResult is error {
+        return credsResult;
+    }
+
+    return credsResult;
```

Comment:
```suggestion
    return check credClient->/(
        headers = {
            [X_AWS_EC2_METADATA_TOKEN]: token
        }
    );
```

Reply:
This throws the error 
```
error: incompatible typedesc aws.s3:IAMCredentials found for 'text/plain' mime type
```

Reply:
As per this error the API might be returning the credentials as a text payload. In that case getting the payload as a string should work.
```ballerina
string credResult = check credClient->/(
    headers = {
        [[X_AWS_EC2_METADATA_TOKEN]: token]
    }
);
return credResult.fromJsonStringWithType();
```

@TharmiganK correct me if I am wrong.

Reply:
This works.

Reply:
Yes, the error indicates that the received response is text and we are trying to bind it to the `IAMCredentials` record. 

But how did it work for the first time? When you are trying to get the Json payload in the current implementation it worked right?

May be the url is wrong, can you try this:
```ballerina
string credResult = check credClient->get("",
    headers = {
        [X_AWS_EC2_METADATA_TOKEN]: token
    }
);
return credResult.fromJsonStringWithType();
```